[#15] 게시물 생성

cyj199637 opened the pull request:
- 게시물과 관련된 기능을 담당하는 PostController, PostService, PostMapper 추가
- 게시물 생성 기능 추가
- FileService 여러 개의 파일을 업로드할 수 있는 uploadFiles 추가
- ResponseEntity 객체를 Util 클래스로 분리

cyj199637 committed:
[#15] 게시물 생성 기능 추가

- 게시물과 관련된 기능을 담당하는 PostController, PostService, PostMapper 추가
- 게시물 생성 기능 추가
- FileService 여러 개의 파일을 업로드할 수 있는 uploadFiles 추가
- ResponseEntity 객체를 Util 클래스로 분리

cyj199637 committed:
Merge branch 'feature/13' into feature/15

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/PostController.java:
```diff
@@ -0,0 +1,36 @@
+
+    private PostService postService;
+
+    @Autowired
```
1. 이 어노테이션을 꼭 달아줘야할까요?
2. 필요없다면 lombok으로 대체할 수 있을 것 같네요~

cyj199637 replied:
생성자 주입을 하면 코드가 복잡해 보일 수 있기 때문에 Lombok에서 `@RequiredArgsConstructor` 를 제공한다.
`@RequiredArgsConstructor` 는 final이나 `@NonNull` 이 설정된 필드를 파라미터로 받는 생성자를 만들어 준다.

f-lab-dev commented on src/main/java/me/liiot/snsserver/controller/UserController.java:
```diff
@@ -27,17 +29,16 @@
     private LoginService loginService;
 
+    @Autowired
+    public UserController(UserService userService, LoginService loginService) {
```
생성자 주입으로 바꿔주신 점 좋네요~ 그럼 멤버변수들도 final을 붙여주는건 어떨까요? 그리고 왜 final을 붙여주는게 좋을까요??

cyj199637 replied:
변수에 final을 선언하면 해당 변수는 선언과 함께 값을 지정해야만 하고 그 뒤로는 값을 바꿀 수 없게 된다. 즉, 읽기만 가능한 변수가 되는 것이다. 그럼에도 불구하고 final을 사용하는 이유는 다음과 같다.
1. static 키워드와 함께 사용하면 모든 클래스에서 사용할 수 있는 상수를 만들 수 있기 때문에 메모리를 아낄 수 있다.
2. 모든 멤버 변수를 final로 선언하면 immutable object를 생성할 수 있다.
2. 값이 변경되지 않는다는 것이 보장되기 때문에 변수를 추적할 필요가 없어져 디버깅하기가 쉬워진다.

f-lab-dev commented on src/main/resources/mapper/PostMapper.xml:
```diff
@@ -0,0 +1,33 @@
+    <insert id="insertImages" parameterType="java.util.List">
+        INSERT INTO image(postId, imageName, imagePath, seq)
+        VALUES
+        <foreach collection="imageUploadInfos" item="info" separator=",">
```
오 벌크 insert를 사용하셨네요. 왜 이게 효율적일까요?

cyj199637 replied:
행을 insert하는데 드는 시간은 다음 요소에 의해 정해지며, 괄호에 적힌 숫자들이 각 요소에 대한 비율을 나타낸다.

- 연결 : (3)
- 서버로 쿼리 전송 : (2)
- 쿼리 구문 분석 : (2)
- 행 삽입 : (1 × 행 크기)
- 색인 삽입 : (1 × 색인 수)
- 마감 : (1)

만약 여러 행을 단일 insert 쿼리로 보낸다면 위와 같은 일련의 프로세스를 행의 수만큼 반복하지만
여러 행을 하나의 bulk insert 쿼리로 보낸다면 행 삽입과 색인 삽입을 제외한 나머지 요소들은
최초 한 번만 수행되기 때문에 훨씬 효율적이다.

f-lab-dev commented on src/main/resources/mapper/PostMapper.xml:
```diff
@@ -0,0 +1,33 @@
+        VALUES(#{userId}, #{content})
+    </insert>
+
+    <select id="getPostId" parameterType="String" resultType="int">
```
해당 유저의 최신글을 불러오는 것 같은데요. 이름도 목적에 맞게 바꿔주시면 어떨까요?

cyj199637 committed:
[#17] 게시물 조회 기능 구현

- 게시물 조회 기능 구현
- Image 객체를 다루던 기능이나 Sql을 각각 FileService, FileMapper로 분리

cyj199637 committed:
[#15] 피드백 반영

- 파일객체를 다루던 기능이나 Sql을 각각 FileService, FileMapper로 분리
- '@RequiredArgsConstructor'를 활용하여 controller, service 객체 생성자코드 제거
- controller, service 객체의 멤버변수에 final 추가

cyj199637 committed:
Merge branch 'feature/15' into feature/17

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/AwsFileService.java:
```diff
@@ -108,6 +108,34 @@ public FileInfo uploadFile(MultipartFile file, String userId) throws FileUploadE
+    public void uploadImages(int postId, List<FileInfo> fileInfos) {
+        List<ImageUploadInfo> imageUploadInfos = new ArrayList<>();
+        int fileInfosLen = fileInfos.size();
+        for (int i=0; i<fileInfosLen; i++) {
```
이거는 stream 연산을 사용해보시면 어떨까요? 그럼 엄청 간결해질 것 같네요~

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/AwsFileService.java:
```diff
@@ -108,6 +108,34 @@ public FileInfo uploadFile(MultipartFile file, String userId) throws FileUploadE
 
+    @Override
+    public void uploadImage(int postId, FileInfo fileInfo) {
+        ImageUploadInfo imageUploadInfo =
```
이렇게 변환해주는 로직이 아래와 중복되는데 이거는 private 메소드로 분리해주면 어떨까요?

f-lab-dev commented on src/main/resources/mapper/FileMapper.xml:
```diff
@@ -0,0 +1,22 @@
+    </insert>
+
+    <insert id="insertImages" parameterType="java.util.List">
+        INSERT INTO image(postId, imageName, imagePath, seq)
```
기존에 리뷰드린 내용이지만 왜 벌크 insert가 빠를까는 답을 갖고계시면 좋을 것 같아요~

cyj199637 committed:
[#15] 피드백 반영

- uploadFiles, uploadImages stream 연산을 활용하여 로직 수정
- uploadImage, uploadImages에 중복된 로직 분리
- uploadFile, uploadFiles에 중복된 로직 분리
- FileUtil의 changeFileName, changeFileNames에 중복된 로직 분리

cyj199637 committed:
Merge branch 'feature/15' into feature/17

cyj199637 committed:
[#19] 사용자 피드 조회

- 사용자 피드 조회 기능 추가
- PostServiceImple의 getPost, getPostsByUser의 중복된 로직 분리

f-lab-dev commented on src/main/java/me/liiot/snsserver/service/AwsFileService.java:
```diff
@@ -42,94 +42,39 @@
-            }
+        HashMap<String, String> newFileNames = FileUtil.changeFileNames(files);
+
+        for (MultipartFile file : files) {
```
이것도 stream 연산으로 바꿀 수 있을 것 같네요~

f-lab-dev reviewed (commented):
대략 보니 foreach와 stream 연산을 헷갈리신 것 같습니다~
자바8에 추가된 스트림 연산을 이용해주세요~

cyj199637 committed:
[#15] 피드백 반영

- stream 연산을 적용하여 로직 수정

cyj199637 committed:
Merge branch 'feature/15' into feature/17

cyj199637 committed:
Merge branch 'feature/17' into feature/19

cyj199637 committed:
[#19] 피드백 반영

- Stream 연산을 적용하여 로직 수정

cyj199637 committed:
[#21] 게시물 수정/삭제 기능 구현

- 게시물 수정 기능 구현
- 게시물 삭제 기능 구현

cyj199637 committed:
[#19] 피드백 반영

- JOIN을 사용하여 불필요한 쿼리 연산 삭제

cyj199637 committed:
[#17] PostServiceImpl의 getPost() 수정

- 불필요한 쿼리를 수행하지 않도록 로직 수정

cyj199637 committed:
Merge branch 'feature/17' into feature/19

cyj199637 committed:
[#21] 피드백 반영

- PostController의 updatePost의 요청 매핑 방식을 Patch로 변경

cyj199637 committed:
[#17] 게시물 조회 기능 캐싱 적용

- 게시물 조회 속도 개선을 위한 캐싱 적용

cyj199637 committed:
Merge branch 'feature/17' into feature/23

cyj199637 committed:
Merge branch 'feature/19' into feature/23

cyj199637 committed:
[#23] 게시물 CRUD 캐싱 적용

- 사용자 게시물 피드를 가져올 때 캐시 저장
- 게시물이 생성, 변경, 삭제될 때마다 해당 사용자의 게시물 피드 캐시 삭제

cyj199637 committed:
[#23] 피드백 반영

- post 객체도 캐싱되도록 변경
- 캐시 유형에 따라 만료 시간을 다르게 설정
- getUserFeed URL 수정

cyj199637 committed:
[#23] 피드백 반영

- getUserFeed의 query String 변경
- CacheKeys를 CacheNames로 변경

cyj199637 committed:
Merge pull request #24 from f-lab-edu/feature/23

[#23] 게시물 CRUD 캐싱 적용

cyj199637 committed:
Merge pull request #22 from f-lab-edu/feature/21

[#21] 게시물 수정/삭제

cyj199637 committed:
Merge pull request #20 from f-lab-edu/feature/19

[#19] 사용자 피드 조회

cyj199637 committed:
Merge pull request #18 from f-lab-edu/feature/17

[#17] 게시물 조회
